miner/oracle.go

```diff
@@ -0,0 +1,195 @@
+	publishEpoch := targetEpoch - 1
+	atxID, err := o.atxDB.GetNodeAtxIDForEpoch(o.nodeID, publishEpoch)
+	if err != nil {
+		o.log.With().Warning("failed to find ATX ID for node",
+			log.Named("publish_epoch", publishEpoch),
+			log.Named("atx_node_id", o.nodeID),
+			log.Err(err))
+		return nil, fmt.Errorf("get ATX ID: %w", err)
+	}
+
+	atx, err := o.atxDB.GetAtxHeader(atxID)
+	if err != nil {
+		o.log.With().Error("failed to get ATX header",
+			log.Named("publish_epoch", publishEpoch),
+			log.Named("atx_node_id", o.nodeID),
+			log.Err(err))
+		return nil, fmt.Errorf("get ATX header: %w", err)
+	}
+	return atx, nil
+}
+
+// calcEligibilityProofs calculates the eligibility proofs of proposals for the miner in the given epoch
+// and returns the proofs along with the epoch's active set.
+func (o *Oracle) calcEligibilityProofs(weight uint64, epoch types.EpochID, beacon []byte) (map[types.LayerID][]types.VotingEligibilityProof, []types.ATXID, error) {
+	beaconDbgStr := types.BytesToHash(beacon).ShortString()
+	logger := o.log.WithFields(epoch, log.String("beacon", beaconDbgStr), log.Uint64("weight", weight))
+
+	// get the previous epoch's total weight
+	totalWeight, activeSet, err := o.atxDB.GetEpochWeight(epoch)
+	if err != nil {
+		return nil, nil, fmt.Errorf("failed to get epoch %v weight: %w", epoch, err)
+	}
+	if totalWeight == 0 {
+		return nil, nil, errZeroEpochWeight
+	}
+	if len(activeSet) == 0 {
+		return nil, nil, errEmptyActiveSet
+	}
+
+	logger = logger.WithFields(log.Uint64("total_weight", totalWeight))
+	logger.Info("calculating eligibility")
+
+	numEligibleSlots, err := proposals.GetNumEligibleSlots(weight, totalWeight, o.avgLayerSize, o.layersPerEpoch)
+	if err != nil {
+		logger.With().Error("failed to get number of eligible proposals", log.Err(err))
+		return nil, nil, fmt.Errorf("oracle get num slots: %w", err)
+	}
+
+	eligibilityProofs := map[types.LayerID][]types.VotingEligibilityProof{}
+	for counter := uint32(0); counter < numEligibleSlots; counter++ {
+		message, err := proposals.SerializeVRFMessage(beacon, epoch, counter)
+		if err != nil {
+			logger.With().Panic("failed to serialize VRF msg", log.Err(err))
+		}
+		vrfSig := o.vrfSigner.Sign(message)
+		eligibleLayer := proposals.CalcEligibleLayer(epoch, o.layersPerEpoch, vrfSig)
+		eligibilityProofs[eligibleLayer] = append(eligibilityProofs[eligibleLayer], types.VotingEligibilityProof{
+			J:   counter,
+			Sig: vrfSig,
+		})
+		logger.Debug("signed vrf message, counter: %v, vrfSig: %v, layer: %v",
+			counter, types.BytesToHash(vrfSig).ShortString(), eligibleLayer)
+	}
+
+	logger.With().Info("proposal eligibility calculated",
+		log.Uint32("total_num_slots", numEligibleSlots),
+		log.Int("num_layers_eligible", len(eligibilityProofs)),
+		log.String("layers_to_num_proposals", prettyPrintProofs(eligibilityProofs)))
+	return eligibilityProofs, activeSet, nil
+}
+
+func prettyPrintProofs(proofs map[types.LayerID][]types.VotingEligibilityProof) string {
```

Comment:
this probably not the new code, but we should use zapcore.ObjectEncoder/zapcore.ArrayEncoder to log complex structs efficiently

Reply:
not new code. i left TODO here to improve. thanks for the suggestion